// @langchain/langgraph-cli/dist/cli/utils/analytics.mjs
import os from "node:os";
import { version } from "./version.mjs";
const SUPABASE_PUBLIC_API_KEY = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6Imt6cmxwcG9qaW5wY3l5YWlweG5iIiwicm9sZSI6ImFub24iLCJpYXQiOjE3MTkyNTc1NzksImV4cCI6MjAzNDgzMzU3OX0.kkVOlLz3BxemA5nP-vat3K4qRtrDuO4SwZSR_htcX9c";
const SUPABASE_URL = "https://kzrlppojinpcyyaipxnb.supabase.co";
async function logData(data) {
    try {
        await fetch(`${SUPABASE_URL}/rest/v1/js_logs`, {
            method: "POST",
            headers: {
                "Content-Type": "application/json",
                apikey: SUPABASE_PUBLIC_API_KEY,
                "User-Agent": "Mozilla/5.0",
            },
            body: JSON.stringify(data),
        });
    }
    catch (error) {
        // pass
    }
}
let analyticsPromise = Promise.resolve();
export function withAnalytics(fn, options) {
    if (process.env.LANGGRAPH_CLI_NO_ANALYTICS === "1") {
        return () => void 0;
    }
    return function (actionCommand) {
        analyticsPromise = analyticsPromise.then(() => logData({
            os: os.platform(),
            os_version: os.release(),
            node_version: process.version,
            cli_version: version,
            cli_command: options?.name ?? actionCommand.name(),
            params: fn?.(actionCommand) ?? {},
        }).catch(() => { }));
    };
}
export async function flushAnalytics() {
    await analyticsPromise;
}
